### ISSUE TITLE
rotational equivariance

### ISSUE THREAD

**Issue by Aries-Wu** [MIDA-group/CoMIR#47]
I can't find the model_biodata_mse.pt in models. Do I need to train to get this file?

**Comment by npielawski** [comment]
Hello Aries!

All the pretrained models are available on the [https://github.com/MIDA-group/CoMIR/releases](release page), which contains both the models and result files. No need to train yourself, unless the modalities of your images are different from ours.

Best,
Nicolas

**Comment by Aries-Wu** [comment]
I'm sorry I missed this page, thanks!

**Comment by Aries-Wu** [comment]
I am sorry that I opened this issue again. I have read it many times carefully, but I still don't understand the rotational equivariance in the paper. Why should I add the operation of rotational equivariance in the training process? I have checked many related papers, but I still don't understand, can you please answer it?

**Comment by npielawski** [comment]
Thanks for your interest in our paper! Our way of adding the rotational equivariance is different from the other existing method, so it isn't surprising that the related papers haven't given you the answers you needed.

I'll try to explain to the best of my abilities, it is a good exercise for myself, so feel free to ask any more questions or clarifications you need. Here we go.

1. We can create representations (CoMIRs) of different images we want to align. When those images are aligned, we want a perfect match between the aligned comirs (meaning that the intensities of the aligned comir of image 1 minus comir of image 2 is 0). If our model is not rotationally equivariant, and the input image is rotated (which will happen since they are misaligned), then the neural network will generate comirs that are different depending on the angle. You can see the disks in the following video: https://youtu.be/iN5GlPWFZ_Q If you have some trouble interpreting the video, perhaps you saw this paper: https://github.com/QUVA-Lab/e2cnn#Demo which inspired us for the illustration format.
2. If we rotate the input image, we would like to get the same **content** in the resulting comir, except that it is going to be rotated. Mathematically, for a transformation T (rotation of 25º) and neural network f: <img src="https://render.githubusercontent.com/render/math?math=f(T_{25\circ}(x)) = T_{25\circ}(f(x))"> This means that it should not matter whether you rotate the input or the comir: you get the same comir rotated regardless. This is by definition, if the model is rotationally equivariant, then this formula is true for any angle and input. And vice versa: if the formula is true, then the model is rot. equivariant.
3. Let's imagine we're just using the triplet loss. In this context, we are trying to maximize the similarity between the anchor A and positive example P and maximize the the dissimilarity between the same anchor A and negative sample N (obviously, they are different, so we still want them to be different comirs). If we have a similarity function s (negative mse, or correlation, whatever), we want to maximize s(f(A), f(P)), i.e. the similarity between the comirs of the anchor and positive sample.
4. Now the last step where we combine everything. If f is rotationally equivariant, than the formula in 2. states that f(T(x))=T(f(x)), so that in our loss we can try to maximize s(f(T(A)), T(f(P))), since both comirs f(A) and f(P) are supposed to be the same after training, then their rotated versions should be the same as well. It's an additional constraint.

Additional comments:
1. In the paper we rotate the CoMIRs back to their original position, for "simplicity", symmetry and also generalization, there could be some corner cases with some other transformations. The formula becomes T'(f(T(A))) = f(A), so if you rotate the input image, compute it's comir and rotate it back, it should be exactly the same as if you had never rotated anything.
2. The explanation above only works for a unique transformation (e.g. rotation by 25º and only 25º). If you want to have the C4 equivariance (multiples of 90º) you need to sum up the losses: s(f(T_0(A)), T_0(f(P))) + s(f(T_90(A)), T_90(f(P))) + s(f(T_180(A)), T_180(f(P))) + s(f(T_270(A)), T_270(f(P))). The work around is to sample a random transformation at each step. The same way that when training a neural network you compute the gradients by sampling a mini-batch instead of the full dataset, you save some sweet computational power. So in practice we still use s(f(T(A)), T(f(P))) and for each sample in the batch and each step, we throw a [four-sided die](https://en.wikipedia.org/wiki/Four-sided_die), if it is 0, then T=no rotation this time, if the die show 1, then T=rotation of 90 degrees, if 2, then T=rotation of 180º and if 3, then T=rotation of 270º.
3. We also showed that (surprisingly) using only four angles of 90 degrees like this gives us a near-perfect coverage of all angles.

I really hope I could help you with your understanding. Again, feel free to ask any question.

Best,
Nicolas

**Comment by Aries-Wu** [comment]
Thanks for your detailed answer! I benefit a lot from it. 

Since our downstream task is registration, we need to ensure that the model is rotational equivariance, and this is achieved by combining the loss into one item. Is that right?

I still have a few things I don't understand.

1. I can't understand the sampling of negative samples. At the beginning of the paper, it is mentioned that training can be done with only one image pairs, and if the negative samples are all from this image pairs, do they become positive samples? I don't know where I misunderstood.

2. The premise of CoMIR requires aligned image pairs, and the purpose of registration is to find a transformation for the unaligned image pairs to align them. If so, how is CoMIR applied to the registration? I may have misunderstood.

Thank you very much for your help!

**Comment by npielawski** [comment]
> Since our downstream task is registration, we need to ensure that the model is rotational equivariance, and this is achieved by combining the loss into one item. 
I am not certain I interpret this sentence properly, but yes: since we are doing (rigid) registration, we have to have a model that is rotationally equivariant. We can achieve that by combining the loss with the transformation scheme.

1. Our loss is not exactly like the triplet loss. InfoNCE takes an anchor, a positive sample and N negative samples. You can choose any N you like. Our data is made out of patches of modalities 1 and 2 that are aligned. Imagine that we have 2 modalities and a batch size of 2, such that we get a matrix of 2x2. Each column is a different modality (column 1 is modality 1, column 2 is modality 2, etc.) and each row represents a different patch. Keep in mind that the images in each row are aligned. What do we put in the loss now? We have anchors and positive examples in our matrix but no negatives yet. The negatives must be different patches, that are not registered with the anchor and the positive sample. For the first element of the batch, the anchor and the positive are in the first row, and we take the second row as negative. Later on, we take the second row as the couple (anchor, positive) and use the first row as negatives.
2. We only need registered pairs of images _during training_, this is because the neural network needs to learn the relationship between modalities. When the model is trained, it "knows" how those relationships occur and does not need registered images anymore.

Hope I could help! If 1. is still not clear, I could make an illustration if you want to.

**Comment by Aries-Wu** [comment]
I have understood the above! Thank you for the great help you have given!

**Comment by Aries-Wu** [comment]
Hi:
I'm sorry to disturb you again. Recently, I reread this paper and found a doubt in the loss, so I'd like to ask.

In the sampling and loss part, if a sample X11 of modality 1 is trained, then the aligned sample X21 in modality 2 is a positive sample, and all the other samples x1i and x2i (i ≠ 1) in modality 1 and modality 2 are negative samples. Therefore, shouldn't the interior of the logarithm function in the loss function be 
e^(h(y11,y21))/{\sum_{i≠1} { e^(h(y11,y1i)) }+\sum_{i≠1} { e^(h(y11,y2i)) } }？

Shouldn't the loss function require that the distance between anchor and positive samples be as large as possible and that between anchor and negative samples be as small as possible? In this paper, the numerator is the distance of a pair of aligned samples, and the denominator is the distance of all aligned sample pairs. And in the sampling part, one anchor, one positive and 2n-2 negative.
 I don't quite understand this place, so I ask you for help. Thank you.

**Comment by Aries-Wu** [comment]
Hi:
I'm sorry to disturb you again. Recently, I reread this paper and found a doubt in the loss, so I'd like to ask.

In the sampling and loss part, if a sample X11 of modality 1 is trained, then the aligned sample X21 in modality 2 is a positive sample, and all the other samples x1i and x2i (i ≠ 1) in modality 1 and modality 2 are negative samples. Therefore, shouldn't the interior of the logarithm function in the loss function be 
e^(h(y11,y21))/{\sum_{i≠1} { e^(h(y11,y1i)) }+\sum_{i≠1} { e^(h(y11,y2i)) } }？

Shouldn't the loss function require that the distance between anchor and positive samples be as large as possible and that between anchor and negative samples be as small as possible? In this paper, the numerator is the distance of a pair of aligned samples, and the denominator is the distance of all aligned sample pairs. And in the sampling part, one anchor, one positive and 2n-2 negative.
 I don't quite understand this place, so I ask you for help. Thank you.

**Comment by Aries-Wu** [comment]
![1](https://user-images.githubusercontent.com/74711632/114876425-e440c180-9e30-11eb-95a6-1a9d8860eebb.png)

**Comment by npielawski** [comment]
Hello,

In our loss, the D_neg describes all the pairs of anchor/negative (assuming the critic is symmetric: h(a,b)=h(b,a)).
The loss you describe is valid and more precise than ours, you could even write the sums as j=1 to n, j \neq i.

In your second paragraph you wrote: "distance between anchor and positive samples be as large as possible" where we actually want the similarity to be as large as possible, making the distance a small as possible. The converse is true for the pair anchor/negative (low similarity, high distance).

Do I answer your questions?

**Comment by Aries-Wu** [comment]
Hello，

Thank you very much for your answer. I understand the second paragraph in your answer, but there are still some uncertainties.

Suppose the sample of modality 1 is X11,..., x1n, and the sample of modality 2 is X21,..., X2N, When training sample X11 in modality 1, D_ Neg = {(X11, X22),..., (X11, X2N)}, is this correct?

Although there are 2n-2 negative samples, only n-1 samples are used. Is this correct？
Thanks~

**Comment by npielawski** [comment]
Hello!

Sorry about the delay for answering. I think what you are saying is correct.

**Comment by wxdrizzle** [comment]
Thank you both for the questions and the detailed answers! They solved several questions I got during reading the paper. Also thank you npielawski for your great work!
